### PR TITLE
Add a "division by zero" check in chunked loss handling in kd_losses.py

### PR DESCRIPTION
**Pull request by insop**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (add a "division by zero" check in chunked loss handling in kd_losses.py)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* add a "division by zero" check in chunked loss handling in kd_losses.py (Issue https://github.com/pytorch/torchtune/issues/2225)

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run `pytest -v tests/torchtune/modules/loss/test_kd_losses.py`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by insop** [comment]
@felipemello1 , @joecummings , PTAL

File: torchtune/modules/loss/kd_losses.py [hunk 1]
```diff
@@ -137,4 +137,7 @@ def forward(
                 student_chunk, teacher_chunk, label_chunk, normalize=False
             )
 
+        if torch.sum(mask.view(-1), dim=0) == 0:
```
Comment (joecummings) [inline torchtune/modules/loss/kd_losses.py:140]: Let's pull this out into a reasonably named variable.

File: torchtune/modules/loss/kd_losses.py [hunk 2]
```diff
@@ -137,4 +137,7 @@ def forward(
...
+            return torch.tensor(0.0, device=student_logits[0].device)
```
Comment (joecummings) [inline torchtune/modules/loss/kd_losses.py:141]: In the discussion on #2094, it seems like if we hit this point, there's probably something wrong with the training data. In that case, should we error out here? Or drop a warning? Or let people shoot themselves in the foot?

cc @lindawangg and @ebsmothers

**Review by joecummings — COMMENTED** [review]
The tests look awesome - thanks!

Just a couple questions/comments on code clarity.

File: torchtune/modules/loss/kd_losses.py (hunk 1 above)
Comment (insop) [reply torchtune/modules/loss/kd_losses.py]: updated.

File: torchtune/modules/loss/kd_losses.py (hunk 2 above)
Comment (insop) [reply torchtune/modules/loss/kd_losses.py:144]: I will wait for inputs, and will update #2094 after this PR is resolved.

Comment (insop) [reply torchtune/modules/loss/kd_losses.py:144]: Any inputs?
@lindawangg , @ebsmothers

Comment (ebsmothers) [reply torchtune/modules/loss/kd_losses.py:144]: Thanks for the bump here. Personally I think this is the right way to handle it. Yes, we could error/raise a warning, but honestly this behavior should be pretty obvious in a loss curve as it will clearly just drop to zero on any such iterations (plus I think zero loss would technically be the "correct" loss value if every token is being ignored). No need to overdo it on handling this edge case here imo

File: tests/torchtune/modules/loss/test_kd_losses.py [hunk 3]
```diff
@@ -17,7 +17,7 @@ def random():
 
 
 class TestForwardKLWithChunkedOutputLoss:
-    def test_forward_kl_loss(self):
+    def setup_forward_kl_loss(self, set_all_masks_zero: bool = False):
```
Comment (ebsmothers) [inline tests/torchtune/modules/loss/test_kd_losses.py:20]: nit: I would do something like "ignore_all_tokens" instead of "set_all_masks_zero" as I think it's clearer

**Review by ebsmothers — COMMENTED** [review]
Thanks for the PR! Just one minor comment on naming in the test, otherwise this looks good

File: tests/torchtune/modules/loss/test_kd_losses.py (hunk 3 above)
Comment (insop) [reply tests/torchtune/modules/loss/test_kd_losses.py]: Sounds better, and updated.

**Comment by insop** [comment]
Thank you @ebsmothers , updated the change.

**Review by ebsmothers — APPROVED** [review]
Thanks for making this change! Once CI is green this should be good to land
